### PR TITLE
Add integration test to check inter-frame communication

### PR DESCRIPTION
**Pull request by esanzgar**
This new integration test checks that `PortFinder`, `PortProvider` and
`PortRPC` work in harmony. A change in one could lead to the others no
working as planned.

### CODE REVIEW

**Comment by esanzgar** [comment]
I have created three simpler tests the following three communication channels:
* guest-host
* guest-sidebar (a little bit more complex)
* sidebar-host

**Comment by esanzgar** [comment]
We exchanged comments about the `subclassing` vs using helper functions [here](https://github.com/hypothesis/client/pull/4092/files#r778077255)

Please, let me know if you want me to using helper functions.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+    });
+    const hostRPC = new PortRPC();
+
+    portFinder.discover('host').then(port => {
+      hostRPC.connect(port);
+      hostRPC.call('ping', response => {
+        assert.equal(response, 'pong');
+        done();
+      });
+    });
```
Comment (robertknight) [inline src/shared/test/integration/inter-frame-communication-test.js:68]: The `done` resolver could potentially be replaced by doing the `portFinder.discover` call in the Promise callback:

```js
const done = new Promise(resolve => {
  portFinder.discover('host').then(port => {
    hostRPC.connect(port);
    hostRPC.call('ping', response => {
      assert.equal(response, 'pong');
      resolve();
    });
  });
});

...

return done;
```

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    await delay(10); // add some realism
+
+    // host frame
+    const portProvider = new PortProvider(window.location.origin);
+    portProvider.listen();
+    const guestRPC = new PortRPC();
+
+    // Register RPC method *before* connection
```
Comment (robertknight) [inline src/shared/test/integration/inter-frame-communication-test.js:77]: I suggest to remove these comments, and leave it to the PortRPC docs to explain that correct usage of the class requires calling `on` before `connect`.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  afterEach(() => {
+    destroyables.forEach(instance => instance.destroy());
+  });
+
+  it('enables the communication between guest-host', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
+
+    // guest frame
```
Comment (robertknight) [inline src/shared/test/integration/inter-frame-communication-test.js:55]:
```suggestion
    // Guest frame
```

.. and similar for other comments.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+    return promise;
+  });
+
+  it('enables the communication between guest-sidebar', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
+
+    // guest frame;
+    const portFinder1 = new PortFinder({
```
Comment (robertknight) [inline src/shared/test/integration/inter-frame-communication-test.js:94]:
```suggestion
    const guestPortFinder = new PortFinder({
```

Alternatively, if you wrap each actor in a separate function, the variable names won't conflict.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    // host frame
+    const portProvider = new PortProvider(window.location.origin);
+    portProvider.listen();
+
+    return promise;
+  });
+
+  it('enables the communication between sidebar-host', async () => {
+    let done;
+    const promise = new Promise(resolve => (done = resolve));
```
Comment (robertknight) [inline src/shared/test/integration/inter-frame-communication-test.js:145]: If you are going to use this pattern, I'd suggest to let `done` be the promise as it is a name of a state (similar to `loaded` or `completed`), and call the resolver `resolveDone`. Alternatively, skip the resolver variable and just use the promise callback:

```js
const done = new Promise(resolve => {
   ...
   hostRPC.call('ping', response => { ...; resolve(); });
});
```

**Review by robertknight — APPROVED** [review]
I think the subclassing is OK for now. There are ways it could be abstracted, but the current implementation is easy enough to follow.

I had a suggestion regarding the grouping of the code that represents each of the different frames in the test, namely to split them into separate functions, to make it clear that they are meant to represent activities happening in different frames which run concurrently and don't share any state (except via a message channel).

Generally LGTM though.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+    portFinder.discover('host').then(port => {
+      hostRPC.connect(port);
+      hostRPC.call('ping', response => {
+        assert.equal(response, 'pong');
+        done();
+      });
+    });
+
+    await delay(10); // add some realism
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/shared/test/integration/inter-frame-communication-test.js:70]: `Guest` can be ready before the `Host` frame. If that's the case, `Guest` will polls the `Host`. I was trying to simulate that scenario.

File: src/shared/test/integration/inter-frame-communication-test.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+  beforeEach(() => {
+    destroyables = [];
+  });
+
+  afterEach(() => {
+    destroyables.forEach(instance => instance.destroy());
+  });
+
+  it('enables the communication between guest-host', async () => {
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/shared/test/integration/inter-frame-communication-test.js:51]: Very good idea! Thanks!

File: src/shared/test/integration/inter-frame-communication-test.js (hunk 1 above)
Comment (esanzgar) [reply src/shared/test/integration/inter-frame-communication-test.js]: Great idea! Thanks!

File: src/shared/test/integration/inter-frame-communication-test.js (hunk 2 above)
Comment (esanzgar) [reply src/shared/test/integration/inter-frame-communication-test.js]: 👍🏼

File: src/shared/test/integration/inter-frame-communication-test.js (hunk 4 above)
Comment (esanzgar) [reply src/shared/test/integration/inter-frame-communication-test.js]: I use separate functions.

**Review by esanzgar — COMMENTED** [review]
I incorporated your feedback on the second commit.

**Pull request merged**